File: src/programmers/level2/오픈채팅방.java
package programmers.level2;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class 오픈채팅방 {
    public static void main(String[] args) {
        String[] record = {"Enter uid1234 Muzi", "Enter uid4567 Prodo", "Leave uid1234", "Enter uid1234 Prodo", "Change uid4567 Ryan"};
        solution(record);
    }

    public static String[] solution(String[] record) {
        Map<String, String> user = new HashMap<>();
        List<String> temp = new ArrayList<>();

        for (String str : record) {
            String[] msg = str.split(" ");
            if (msg[0].equals("Enter"))
                user.put(msg[1], msg[2]);
            else if (msg[0].equals("Change"))
                user.put(msg[1], msg[2]);
        }

        for (String str : record) {
            String[] msg = str.split(" ");

            if (msg[0].equals("Enter"))
                temp.add(user.get(msg[1]) + "님이 들어왔습니다.");
            else if (msg[0].equals("Leave"))
                temp.add(user.get(msg[1]) + "님이 나갔습니다.");
        }

        /* answer에 바로 담으면 런타임 에러 리스트에 담고 return 때 변경 >>> why? */

        String[] answer = new String[temp.size()];
        return temp.toArray(answer);
    }
}

File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.js
function solution(record) {
    var answer = [];
    let map = new Map();
    
    record.forEach((action) => {
        let arr = action.split(' ');
        if(arr[0] === 'Enter' || arr[0] === 'Change')
            map.set(arr[1], arr[2]);
    } )
    
    record.forEach((action) => {
        let arr = action.split(' ');
        if(arr[0] === 'Enter')
            answer.push(`${map.get(arr[1])}님이 들어왔습니다.`);
        if(arr[0] === 'Leave')
            answer.push(`${map.get(arr[1])}님이 나갔습니다.`);
    } )
    
    
    return answer;
}